### PR TITLE
Simplify MNIST example.

### PR DESCRIPTION
**Pull request by lukaszlew**
# What does this PR do?
This PR aims to simplify the code and make it more readable.
It is aimed at making it more attractive to ML beginners.

- Remove a duplicate cross entropy calculation in `compute_metrics`.
- Use the same code for train set and test set
- Expose the one liner `update_model`.
- Replace fairly hard to read dictionaries and dictionary comprehensions with variables.
- Remove `jax.device_get` calls.
- Replace `logging.info` with `print` for the main log.
  This improves train log visibility and removes the need for `googlelog.Capture()`.
- Make the training log one line per epoch with constant with numbers making it easy to compare them.
- Make naming uniform (`test`, `eval` into `test`), in line with [PyTorch example](https://github.com/pytorch/examples/blob/master/mnist/main.py).

### CODE REVIEW

File: examples/mnist/train.py [hunk 1]
```diff
@@ -51,38 +51,24 @@ def __call__(self, x):
...
-      'accuracy': accuracy,
-  }
-  return metrics
-
-
 @jax.jit
-def train_step(state, batch):
-  """Train for a single step."""
+def apply_model(state, images, labels):
+  """Compute loss, accurcy and gradients for a single batch."""
```
Comment (andsteing) [inline examples/mnist/train.py:56]: `"""Computes gradients, loss, accuracy and for a single batch."""`

File: examples/mnist/train.py [hunk 2]
```diff
@@ -161,17 +135,19 @@ def train_and_evaluate(config: ml_collections.ConfigDict,
...
+      state, train_ds, config.batch_size, epoch, input_rng)
+    _, test_loss, test_accu = apply_model(
+      state, test_ds['image'], test_ds['label'])
+
+    print(
+      'epoch:% 3d, train_loss: %.4f, train_accu: %.2f, test_loss: %.4f, test_accu: %.2f' %
+      (epoch, train_loss, train_accu * 100, test_loss, test_accu * 100))
+
+    summary_writer.scalar('train_loss', train_loss, epoch)
+    summary_writer.scalar('train_accu', train_accu, epoch)
```
Comment (andsteing) [inline examples/mnist/train.py:148]: Let's keep the original name `train_accuracy`.
(Here and elsewhere)

**Review by andsteing — CHANGES_REQUESTED** [review]
Thanks for simplifying the example, this looks great!

Could you please also do the following in this PR?

1. Re-run the example and attach link to https://tensorboard.dev/ metrics.
2. Make sure the Colab `examples/mnist/mnist.ipynb` still runs with your changes.

File: examples/mnist/train.py (hunk 1 above)
Comment (lukaszlew) [reply examples/mnist/train.py:56]: Done

File: examples/mnist/train.py (hunk 2 above)
Comment (lukaszlew) [reply examples/mnist/train.py:148]: done

**Comment by lukaszlew** [comment]
[tensorboard link](https://tensorboard.dev/experiment/1S90u2cARPm7vmrvN9VXCQ/#scalars&_smoothingWeight=0)

Log from the training:
```
epoch:  1, train_loss: 0.2709, train_accuracy: 91.91, test_loss: 0.0491, test_accuracy: 98.39
epoch:  2, train_loss: 0.0513, train_accuracy: 98.41, test_loss: 0.0375, test_accuracy: 98.72
epoch:  3, train_loss: 0.0382, train_accuracy: 98.79, test_loss: 0.0252, test_accuracy: 99.10
epoch:  4, train_loss: 0.0271, train_accuracy: 99.10, test_loss: 0.0255, test_accuracy: 99.14
epoch:  5, train_loss: 0.0229, train_accuracy: 99.25, test_loss: 0.0270, test_accuracy: 99.13
epoch:  6, train_loss: 0.0180, train_accuracy: 99.39, test_loss: 0.0271, test_accuracy: 99.10
epoch:  7, train_loss: 0.0154, train_accuracy: 99.50, test_loss: 0.0272, test_accuracy: 99.11
epoch:  8, train_loss: 0.0127, train_accuracy: 99.56, test_loss: 0.0295, test_accuracy: 99.10
epoch:  9, train_loss: 0.0105, train_accuracy: 99.66, test_loss: 0.0315, test_accuracy: 99.02
epoch: 10, train_loss: 0.0085, train_accuracy: 99.74, test_loss: 0.0273, test_accuracy: 99.22
```

[Confirmed that the colab still works.](https://colab.research.google.com/drive/1nC6cG8EnJNE3990rIFApJ9rmlOVaeQJI?usp=sharing)

File: examples/mnist/train.py [hunk 3]
```diff
@@ -51,38 +51,24 @@ def __call__(self, x):
...
-      'accuracy': accuracy,
-  }
-  return metrics
-
-
 @jax.jit
-def train_step(state, batch):
-  """Train for a single step."""
+def apply_model(state, images, labels):
+  """Compute gradientsh loss and accuracy for a single batch."""
```
Comment (andsteing) [inline examples/mnist/train.py:56]: `"""Computes gradients, loss, accuracy and for a single batch."""`

**Review by andsteing — CHANGES_REQUESTED** [review]
Typo in updated pydoc, otherwise LGTM.

**Review by andsteing — APPROVED** [review]

File: examples/mnist/train.py (hunk 3 above)
Comment (lukaszlew) [reply examples/mnist/train.py]: done
